### DISCUSSION TITLE
I don't quite understand how `fork` works

### DISCUSSION THREAD

**Discussion by QuarticCat** [ConorWilliams/libfork#25] [Q&A] [answered]
After `fork_awaitable::await_suspend` being called, the parent task will be pushed to WSQ, and then the function continues to `resume` the child task. But before the first suspension of the child task, the state of the parent task has not been changed to the next one yet. While at this timepoint, the parent task can be stolen because it's in WSQ already. If another worker steals the parent task and `resume`s it, the resume function will not start from the state after `co_await fork`. Am I wrong?

**Comment by ConorWilliams** [comment]
I don't fully follow your question but I'll do my best to answer, please ask if you need more clarification:

> After `fork_awaitable::await_suspend` being called, the parent task will be pushed to WSQ, and then the function continues to resume the child task. 

It is the _executing thread_ that resumes the child task.

> But before the first suspension of the child task, the state of the parent task has not been changed to the next one yet. While at this timepoint, the parent task can be stolen because it's in WSQ already. 

In `fork_awaitable::await_suspend` both the parent and the child tasks are suspended. A handle to the parent (which will resume just after the `fork` call) is puched to the executors WSQ.

> If another worker steals the parent task and resumes it, the resume function will not start from the state after co_await fork. Am I wrong?

Yes, you are wrong, the stolen continuation (of the parent) will resume after the `fork`.

**Comment by QuarticCat** [comment]
According to the expansion of cppinsights in [this example](https://cppinsights.io/lnk?code=I2luY2x1ZGUgPGlvc3RyZWFtPgojaW5jbHVkZSA8Y29yb3V0aW5lPgoKc3RydWN0IEF3YWl0YWJsZTogc3RkOjpzdXNwZW5kX2Fsd2F5cyB7CiAgICBzdGQ6OmNvcm91dGluZV9oYW5kbGU8PiBhd2FpdF9zdXNwZW5kKHN0ZDo6Y29yb3V0aW5lX2hhbmRsZTw+KXsKCQkvLyBUT0RPCiAgICB9Cn07CgpzdHJ1Y3QgVGFza3sKICAgIHN0cnVjdCBwcm9taXNlX3R5cGU7CiAgICB1c2luZyBoYW5kbGUgPSBzdGQ6OmNvcm91dGluZV9oYW5kbGU8cHJvbWlzZV90eXBlPjsKICAgIHN0cnVjdCBwcm9taXNlX3R5cGV7CiAgICAgICAgYXV0byBpbml0aWFsX3N1c3BlbmQoKXsKICAgICAgICAgICAgcmV0dXJuIHN0ZDo6c3VzcGVuZF9hbHdheXN7fTsKICAgICAgICB9CiAgICAgICAgYXV0byBmaW5hbF9zdXNwZW5kKCkgbm9leGNlcHR7CiAgICAgICAgICAgIHJldHVybiBzdGQ6OnN1c3BlbmRfYWx3YXlze307CiAgICAgICAgfQogICAgICAgIHZvaWQgdW5oYW5kbGVkX2V4Y2VwdGlvbigpe30KICAgICAgICB2b2lkIHJldHVybl92b2lkKCl7fQogICAgICAgIGF1dG8gZ2V0X3JldHVybl9vYmplY3QoKXsKICAgICAgICAgICByZXR1cm4gVGFza3toYW5kbGU6OmZyb21fcHJvbWlzZSgqdGhpcyl9OwogICAgICAgIH0KICAgICAgCUF3YWl0YWJsZSBhd2FpdF90cmFuc2Zvcm0oQXdhaXRhYmxlKSB7CiAgICAgICAgCS8vIFRPRE8KICAgICAgICB9CiAgICB9OwogICAgaGFuZGxlIGNvcm87Cn07CgpUYXNrIGZ1bigpewogICAgc3RkOjpjb3V0PDwiYmVnaW5cbiI7CiAgICBjb19hd2FpdCBBd2FpdGFibGV7fTsKICAgIHN0ZDo6Y291dDw8ImVuZFxuIjsKfQo=&insightsOptions=cpp20,use-libcpp,edu-show-coroutines&std=cpp20&rev=1.0), invocation of `await_suspend` is before the assignment of `__suspend_index`. So just after `tls::context()->push`, the `__suspend_index` is still unchanged, while at this timepoint another thread can steal and resume the parent.

**Reply by ConorWilliams** [reply, marked as answer]
[Cpp reference](https://en.cppreference.com/w/cpp/language/coroutines) says:

> 
> Then, awaiter.await_ready() is called (this is a short-cut to avoid the cost of suspension if it's known that the result is ready or can be completed synchronously). If its result, contextually-converted to bool is false then
> 
> The coroutine is suspended (its coroutine state is populated with local variables and current suspension point).
> awaiter.await_suspend(handle) is called, where handle is the coroutine handle representing the current coroutine. Inside that function, the suspended coroutine state is observable via that handle, and it's this function's responsibility to schedule it to resume on some executor, or to be destroyed (returning false counts as scheduling)

I expect there could be several reasons for the the observation you see:
- The index is being set for the next suspend point not the current one.
- Optimisations are obfuscating your code, a smart compiler could see that nothing is done with the handle and that it safe to defer the increment (or not do it at all)
- Your code contains multiple UB: non-void functions must terminate with a return.

**Reply by QuarticCat** [reply]
That makes sense. Thank you for you answers!
